### PR TITLE
tests: counter: counter_basic_api: Rework test to use struct device

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

**Comment by galak** [comment]
marked DNM as need to under the 'RTC_0" bits.

**Comment by galak** [comment]
@hakehuang what was the intent of commit 3d0767c36e6.  Curious why only `RTC_0` and for which SoCs does this apply?

**Comment by hakehuang** [comment]
> @hakehuang what was the intent of commit [3d0767c](https://github.com/zephyrproject-rtos/zephyr/commit/3d0767c36e6bebf6b822b2c5298ea97e34007c68). Curious why only `RTC_0` and for which SoCs does this apply?

@galak , the counter need special setting for RTC_0 see #82268629918272118559a8c200583a2af3d1ca38 I just move the changes to RTC_0 only and use default for other counter.

**Comment by galak** [comment]
> @galak , the counter need special setting for RTC_0 see [8226862](https://github.com/zephyrproject-rtos/zephyr/commit/82268629918272118559a8c200583a2af3d1ca38) I just move the changes to RTC_0 only and use default for other counter.

@hakehuang any reason why this only applies to RTC_0?

**Comment by hakehuang** [comment]
> @hakehuang any reason why this only applies to RTC_0?

only RTC_0 impacted platfrom has sepcial requirements see code below

```
#if defined(CONFIG_COUNTER_MCUX_RTC) || defined(CONFIG_COUNTER_RTC_STM32) || \
	defined(CONFIG_COUNTER_MCUX_LPC_RTC)
#define COUNTER_PERIOD_US_VAL (USEC_PER_SEC * 2U)
#else
#define COUNTER_PERIOD_US_VAL 20000
#endif
```

**Comment by galak** [comment]
> 

@hakehuang but wouldn't this apply equally to RTC_1 on those platforms?

**Comment by hakehuang** [comment]
> > 
> 
> @hakehuang but wouldn't this apply equally to RTC_1 on those platforms?

yes, but for intial test we only touch RTC_0, if RTC_1 is also required, then we need some changes

**Comment by galak** [comment]
@nordic-krch ping.

**Review by nordic-krch — DISMISSED** [review]

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 1]
```diff
@@ -19,114 +19,127 @@ static void top_handler(const struct device *dev, void *user_data);
...
-#define INST_DT_COMPAT_LABEL(n, compat) DT_LABEL(DT_INST(n, compat))
-/* Generate a list of LABELs for all instances of the "compat" */
-#define LABELS_FOR_DT_COMPAT(compat) \
+#define INST_DT_COMPAT_DEV(n, compat) DEVICE_DT_GET(DT_INST(n, compat))
+/* Generate a list of devices for all instances of the "compat" */
+#define DEVS_FOR_DT_COMPAT(compat) \
 	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat), \
 		   (LISTIFY(DT_NUM_INST_STATUS_OKAY(compat), \
-			    INST_DT_COMPAT_LABEL, (,), compat),), ())
+			    INST_DT_COMPAT_DEV, (,), compat),), ())
```
Comment (gmarull) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:30]:
```c
#define DEVICE_DT_GET_AND_COMMA(node_id) DEVICE_DT_GET(node_id),
```

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 2]
```diff
@@ -19,114 +19,127 @@ static void top_handler(const struct device *dev, void *user_data);
...
-	/* NOTE: there is no trailing comma, as the DT_LABELS_FOR_COMPAT
+	/* NOTE: there is no trailing comma, as the DEVS_FOR_DT_COMPAT
 	 * handles it.
 	 */
-	LABELS_FOR_DT_COMPAT(arm_cmsdk_timer)
-	LABELS_FOR_DT_COMPAT(arm_cmsdk_dtimer)
-	LABELS_FOR_DT_COMPAT(microchip_xec_timer)
-	LABELS_FOR_DT_COMPAT(nxp_imx_epit)
-	LABELS_FOR_DT_COMPAT(nxp_imx_gpt)
+	DEVS_FOR_DT_COMPAT(arm_cmsdk_timer)
```
Comment (gmarull) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:71]:
```suggestion
	DT_FOREACH_STATUS_OKAY(arm_cmsdk_timer, DEVICE_DT_GET_AND_COMMA)
```

same for others

**Review by gmarull — CHANGES_REQUESTED** [review]

**Review by gmarull — APPROVED** [review]

**Comment by galak** [comment]
@nordic-krch ping, one more time.

**Comment by galak** [comment]
Going to ignore the doc failure which is a CI issue, and not one with the PR.

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**